### PR TITLE
test: deflake by making registerTestServerPorts thread safe

### PR DESCRIPTION
**Pull request by lambdai**
Signed-off-by: Yuchen Dai <silentdai@gmail.com>
Description: 
The is data race, reported by tsan
The root cause is  listener manager is accessed in both envoy master thread and test thread.
Use Dispatcher::post to access ListenerManager::listeners()

Risk Level:
Testing:
Docs Changes:
Release Notes:
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by lambdai** [comment]
The motivation of posting `listeners()` to envoy's master thread.
```
2020-03-25T09:30:32.9629093Z WARNING: ThreadSanitizer: data race (pid=17)
2020-03-25T09:30:32.9629493Z   Write of size 8 at 0x7b480001ffa8 by thread T16:
2020-03-25T09:30:32.9630929Z     #0 std::__1::list<std::__1::unique_ptr<Envoy::Server::ListenerImpl, std::__1::default_delete<Envoy::Server::ListenerImpl> >, std::__1::allocator<std::__1::unique_ptr<Envoy::Server::ListenerImpl, std::__1::default_delete<Envoy::Server::ListenerImpl> > > >::__link_nodes_at_back(std::__1::__list_node_base<std::__1::unique_ptr<Envoy::Server::ListenerImpl, std::__1::default_delete<Envoy::Server::ListenerImpl> >, void*>*, std::__1::__list_node_base<std::__1::unique_ptr<Envoy::Server::ListenerImpl, std::__1::default_delete<Envoy::Server::ListenerImpl> >, void*>*) /opt/llvm/bin/../include/c++/v1/list:1191:27 (listener_lds_integration_test+0x2a32851)
2020-03-25T09:30:32.9634154Z     #1 void std::__1::list<std::__1::unique_ptr<Envoy::Server::ListenerImpl, std::__1::default_delete<Envoy::Server::ListenerImpl> >, std::__1::allocator<std::__1::unique_ptr<Envoy::Server::ListenerImpl, std::__1::default_delete<Envoy::Server::ListenerImpl> > > >::emplace_back<std::__1::unique_ptr<Envoy::Server::ListenerImpl, std::__1::default_delete<Envoy::Server::ListenerImpl> > >(std::__1::unique_ptr<Envoy::Server::ListenerImpl, std::__1::default_delete<Envoy::Server::ListenerImpl> >&&) /opt/llvm/bin/../include/c++/v1/list:1682:5 (listener_lds_integration_test+0x29f09ff)
2020-03-25T09:30:32.9636553Z     #2 Envoy::Server::ListenerManagerImpl::addOrUpdateListenerInternal(envoy::config::listener::v3::Listener const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) /proc/self/cwd/source/server/listener_manager_impl.cc:474:25 (listener_lds_integration_test+0x29cb12f)
2020-03-25T09:30:32.9638368Z     #3 Envoy::Server::ListenerManagerImpl::addOrUpdateListener(envoy::config::listener::v3::Listener const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool) /proc/self/cwd/source/server/listener_manager_impl.cc:344:12 (listener_lds_integration_test+0x29c91ce)
2020-03-25T09:30:32.9640199Z     #4 Envoy::Server::LdsApiImpl::onConfigUpdate(google::protobuf::RepeatedPtrField<envoy::service::discovery::v3::Resource> const&, google::protobuf::RepeatedPtrField<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) /proc/self/cwd/source/server/lds_api.cc:73:29 (listener_lds_integration_test+0x2b0931a)
2020-03-25T09:30:32.9641968Z     #5 Envoy::Server::LdsApiImpl::onConfigUpdate(google::protobuf::RepeatedPtrField<google::protobuf::Any> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) /proc/self/cwd/source/server/lds_api.cc:126:3 (listener_lds_integration_test+0x2b09e57)
2020-03-25T09:30:32.9643971Z     #6 non-virtual thunk to Envoy::Server::LdsApiImpl::onConfigUpdate(google::protobuf::RepeatedPtrField<google::protobuf::Any> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) /proc/self/cwd/source/server/lds_api.cc (listener_lds_integration_test+0x2b09f17)
2020-03-25T09:30:32.9645528Z     #7 Envoy::Config::GrpcSubscriptionImpl::onConfigUpdate(google::protobuf::RepeatedPtrField<google::protobuf::Any> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) /proc/self/cwd/source/common/config/grpc_subscription_impl.cc:62:14 (listener_lds_integration_test+0x49403bb)
2020-03-25T09:30:32.9647657Z     #8 non-virtual thunk to Envoy::Config::GrpcSubscriptionImpl::onConfigUpdate(google::protobuf::RepeatedPtrField<google::protobuf::Any> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) /proc/self/cwd/source/common/config/grpc_subscription_impl.cc (listener_lds_integration_test+0x4940747)
2020-03-25T09:30:32.9649264Z     #9 Envoy::Config::GrpcMuxImpl::onDiscoveryResponse(std::__1::unique_ptr<envoy::service::discovery::v3::DiscoveryResponse, std::__1::default_delete<envoy::service::discovery::v3::DiscoveryResponse> >&&) /proc/self/cwd/source/common/config/grpc_mux_impl.cc:176:27 (listener_lds_integration_test+0x4946885)
2020-03-25T09:30:32.9651147Z     #10 non-virtual thunk to Envoy::Config::GrpcMuxImpl::onDiscoveryResponse(std::__1::unique_ptr<envoy::service::discovery::v3::DiscoveryResponse, std::__1::default_delete<envoy::service::discovery::v3::DiscoveryResponse> >&&) /proc/self/cwd/source/common/config/grpc_mux_impl.cc (listener_lds_integration_test+0x4946f4f)
2020-03-25T09:30:32.9653667Z     #11 Envoy::Config::GrpcStream<envoy::service::discovery::v3::DiscoveryRequest, envoy::service::discovery::v3::DiscoveryResponse>::onReceiveMessage(std::__1::unique_ptr<envoy::service::discovery::v3::DiscoveryResponse, std::__1::default_delete<envoy::service::discovery::v3::DiscoveryResponse> >&&) /proc/self/cwd/bazel-out/k8-dbg/bin/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:83:17 (listener_lds_integration_test+0x2c6f4da)
2020-03-25T09:30:32.9655829Z     #12 Envoy::Grpc::AsyncStreamCallbacks<envoy::service::discovery::v3::DiscoveryResponse>::onReceiveMessageRaw(std::__1::unique_ptr<Envoy::Buffer::Instance, std::__1::default_delete<Envoy::Buffer::Instance> >&&) /proc/self/cwd/bazel-out/k8-dbg/bin/source/common/grpc/_virtual_includes/typed_async_client_lib/common/grpc/typed_async_client.h:90:5 (listener_lds_integration_test+0x2c6f166)
2020-03-25T09:30:32.9657149Z     #13 Envoy::Grpc::GoogleAsyncStreamImpl::handleOpCompletion(Envoy::Grpc::GoogleAsyncTag::Operation, bool) /proc/self/cwd/source/common/grpc/google_async_client_impl.cc:328:32 (listener_lds_integration_test+0x2da8b75)
2020-03-25T09:30:32.9658043Z     #14 Envoy::Grpc::GoogleAsyncStreamImpl::onCompletedOps() /proc/self/cwd/source/common/grpc/google_async_client_impl.cc:258:5 (listener_lds_integration_test+0x2da4846)
2020-03-25T09:30:32.9658963Z     #15 Envoy::Grpc::GoogleAsyncClientThreadLocal::completionThread()::$_1::operator()() const /proc/self/cwd/source/common/grpc/google_async_client_impl.cc:64:50 (listener_lds_integration_test+0x2daff0b)
2020-03-25T09:30:32.9660156Z     #16 decltype(std::__1::forward<Envoy::Grpc::GoogleAsyncClientThreadLocal::completionThread()::$_1&>(fp)()) std::__1::__invoke<Envoy::Grpc::GoogleAsyncClientThreadLocal::completionThread()::$_1&>(Envoy::Grpc::GoogleAsyncClientThreadLocal::completionThread()::$_1&) /opt/llvm/bin/../include/c++/v1/type_traits:3530:1 (listener_lds_integration_test+0x2dafe60)
2020-03-25T09:30:32.9661400Z     #17 void std::__1::__invoke_void_return_wrapper<void>::__call<Envoy::Grpc::GoogleAsyncClientThreadLocal::completionThread()::$_1&>(Envoy::Grpc::GoogleAsyncClientThreadLocal::completionThread()::$_1&) /opt/llvm/bin/../include/c++/v1/__functional_base:348:9 (listener_lds_integration_test+0x2dafdc0)
2020-03-25T09:30:32.9662778Z     #18 std::__1::__function::__alloc_func<Envoy::Grpc::GoogleAsyncClientThreadLocal::completionThread()::$_1, std::__1::allocator<Envoy::Grpc::GoogleAsyncClientThreadLocal::completionThread()::$_1>, void ()>::operator()() /opt/llvm/bin/../include/c++/v1/functional:1533:16 (listener_lds_integration_test+0x2dafd60)
2020-03-25T09:30:32.9664124Z     #19 std::__1::__function::__func<Envoy::Grpc::GoogleAsyncClientThreadLocal::completionThread()::$_1, std::__1::allocator<Envoy::Grpc::GoogleAsyncClientThreadLocal::completionThread()::$_1>, void ()>::operator()() /opt/llvm/bin/../include/c++/v1/functional:1707:12 (listener_lds_integration_test+0x2dae02f)
2020-03-25T09:30:32.9665233Z     #20 std::__1::__function::__value_func<void ()>::operator()() const /opt/llvm/bin/../include/c++/v1/functional:1860:16 (listener_lds_integration_test+0x25c6826)
2020-03-25T09:30:32.9666272Z     #21 std::__1::function<void ()>::operator()() const /opt/llvm/bin/../include/c++/v1/functional:2419:12 (listener_lds_integration_test+0x25c64b8)
2020-03-25T09:30:32.9667130Z     #22 Envoy::Event::DispatcherImpl::runPostCallbacks() /proc/self/cwd/source/common/event/dispatcher_impl.cc:225:5 (listener_lds_integration_test+0x4f2441f)
2020-03-25T09:30:32.9668372Z     #23 Envoy::Event::DispatcherImpl::DispatcherImpl(std::__1::unique_ptr<Envoy::Buffer::WatermarkFactory, std::__1::default_delete<Envoy::Buffer::WatermarkFactory> >&&, Envoy::Api::Api&, Envoy::Event::TimeSystem&)::$_1::operator()() const /proc/self/cwd/source/common/event/dispatcher_impl.cc:43:58 (listener_lds_integration_test+0x4f2954b)
2020-03-25T09:30:32.9671211Z     #24 decltype(std::__1::forward<Envoy::Event::DispatcherImpl::DispatcherImpl(std::__1::unique_ptr<Envoy::Buffer::WatermarkFactory, std::__1::default_delete<Envoy::Buffer::WatermarkFactory> >&&, Envoy::Api::Api&, Envoy::Event::TimeSystem&)::$_1&>(fp)()) std::__1::__invoke<Envoy::Event::DispatcherImpl::DispatcherImpl(std::__1::unique_ptr<Envoy::Buffer::WatermarkFactory, std::__1::default_delete<Envoy::Buffer::WatermarkFactory> >&&, Envoy::Api::Api&, Envoy::Event::TimeSystem&)::$_1&>(Envoy::Event::DispatcherImpl::DispatcherImpl(std::__1::unique_ptr<Envoy::Buffer::WatermarkFactory, std::__1::default_delete<Envoy::Buffer::WatermarkFactory> >&&, Envoy::Api::Api&, Envoy::Event::TimeSystem&)::$_1&) /opt/llvm/bin/../include/c++/v1/type_traits:3530:1 (listener_lds_integration_test+0x4f294a0)
2020-03-25T09:30:32.9675877Z     #25 void std::__1::__invoke_void_return_wrapper<void>::__call<Envoy::Event::DispatcherImpl::DispatcherImpl(std::__1::unique_ptr<Envoy::Buffer::WatermarkFactory, std::__1::default_delete<Envoy::Buffer::WatermarkFactory> >&&, Envoy::Api::Api&, Envoy::Event::TimeSystem&)::$_1&>(Envoy::Event::DispatcherImpl::DispatcherImpl(std::__1::unique_ptr<Envoy::Buffer::WatermarkFactory, std::__1::default_delete<Envoy::Buffer::WatermarkFactory> >&&, Envoy::Api::Api&, Envoy::Event::TimeSystem&)::$_1&) /opt/llvm/bin/../include/c++/v1/__functional_base:348:9 (listener_lds_integration_test+0x4f29400)
2020-03-25T09:30:33.5112508Z     #26 std::__1::__function::__alloc_func<Envoy::Event::DispatcherImpl::DispatcherImpl(std::__1::unique_ptr<Envoy::Buffer::WatermarkFactory, std::__1::default_delete<Envoy::Buffer::WatermarkFactory> >&&, Envoy::Api::Api&, Envoy::Event::TimeSystem&)::$_1, std::__1::allocator<Envoy::Event::DispatcherImpl::DispatcherImpl(std::__1::unique_ptr<Envoy::Buffer::WatermarkFactory, std::__1::default_delete<Envoy::Buffer::WatermarkFactory> >&&, Envoy::Api::Api&, Envoy::Event::TimeSystem&)::$_1>, void ()>::operator()() /opt/llvm/bin/../include/c++/v1/functional:1533:16 (listener_lds_integration_test+0x4f293a0)
2020-03-25T09:30:33.5116339Z     #27 std::__1::__function::__func<Envoy::Event::DispatcherImpl::DispatcherImpl(std::__1::unique_ptr<Envoy::Buffer::WatermarkFactory, std::__1::default_delete<Envoy::Buffer::WatermarkFactory> >&&, Envoy::Api::Api&, Envoy::Event::TimeSystem&)::$_1, std::__1::allocator<Envoy::Event::DispatcherImpl::DispatcherImpl(std::__1::unique_ptr<Envoy::Buffer::WatermarkFactory, std::__1::default_delete<Envoy::Buffer::WatermarkFactory> >&&, Envoy::Api::Api&, Envoy::Event::TimeSystem&)::$_1>, void ()>::operator()() /opt/llvm/bin/../include/c++/v1/functional:1707:12 (listener_lds_integration_test+0x4f2766f)
2020-03-25T09:30:33.5118436Z     #28 std::__1::__function::__value_func<void ()>::operator()() const /opt/llvm/bin/../include/c++/v1/functional:1860:16 (listener_lds_integration_test+0x25c6826)
2020-03-25T09:30:33.5119441Z     #29 std::__1::function<void ()>::operator()() const /opt/llvm/bin/../include/c++/v1/functional:2419:12 (listener_lds_integration_test+0x25c64b8)
2020-03-25T09:30:33.5120802Z     #30 Envoy::Event::TimerImpl::TimerImpl(Envoy::CSmartPtr<event_base, &(event_base_free)>&, std::__1::function<void ()>, Envoy::Event::Dispatcher&)::$_0::operator()(int, short, void*) const /proc/self/cwd/source/common/event/timer_impl.cc:15:3 (listener_lds_integration_test+0x50b4515)
2020-03-25T09:30:33.5122381Z     #31 Envoy::Event::TimerImpl::TimerImpl(Envoy::CSmartPtr<event_base, &(event_base_free)>&, std::__1::function<void ()>, Envoy::Event::Dispatcher&)::$_0::__invoke(int, short, void*) /proc/self/cwd/source/common/event/timer_impl.cc:15:3 (listener_lds_integration_test+0x50b4476)
2020-03-25T09:30:33.5123614Z     #32 event_process_active_single_queue /b/f/w/external/com_github_libevent_libevent/event.c:1713:4 (listener_lds_integration_test+0x5689d55)
2020-03-25T09:30:33.5124436Z     #33 event_process_active /b/f/w/external/com_github_libevent_libevent/event.c:1805:9 (listener_lds_integration_test+0x567f3a8)
2020-03-25T09:30:33.5125242Z     #34 event_base_loop /b/f/w/external/com_github_libevent_libevent/event.c:2047:12 (listener_lds_integration_test+0x567d073)
2020-03-25T09:30:33.5126181Z     #35 Envoy::Event::LibeventScheduler::run(Envoy::Event::Dispatcher::RunType) /proc/self/cwd/source/common/event/libevent_scheduler.cc:47:3 (listener_lds_integration_test+0x50b21ef)
2020-03-25T09:30:33.5127277Z     #36 Envoy::Event::DispatcherImpl::run(Envoy::Event::Dispatcher::RunType) /proc/self/cwd/source/common/event/dispatcher_impl.cc:197:19 (listener_lds_integration_test+0x4f242f1)
2020-03-25T09:30:33.5128405Z     #37 Envoy::Server::InstanceImpl::run() /proc/self/cwd/source/server/server.cc:556:16 (listener_lds_integration_test+0x28f12cb)
2020-03-25T09:30:33.5130036Z     #38 Envoy::IntegrationTestServerImpl::createAndRunEnvoyServer(Envoy::OptionsImpl&, Envoy::Event::TimeSystem&, std::__1::shared_ptr<Envoy::Network::Address::Instance const>, Envoy::ListenerHooks&, Envoy::Thread::BasicLockable&, Envoy::Server::ComponentFactory&, std::__1::unique_ptr<Envoy::Runtime::RandomGenerator, std::__1::default_delete<Envoy::Runtime::RandomGenerator> >&&, absl::optional<std::__1::reference_wrapper<Envoy::ProcessObject> >) /proc/self/cwd/test/integration/server.cc:213:12 (listener_lds_integration_test+0x26eba80)
2020-03-25T09:30:33.5132021Z     #39 Envoy::IntegrationTestServer::threadRoutine(Envoy::Network::Address::IpVersion, bool, absl::optional<std::__1::reference_wrapper<Envoy::ProcessObject> >, bool, bool, unsigned int) /proc/self/cwd/test/integration/server.cc:182:3 (listener_lds_integration_test+0x26eb4d2)
2020-03-25T09:30:33.5133600Z     #40 Envoy::IntegrationTestServer::start(Envoy::Network::Address::IpVersion, std::__1::function<void ()>, bool, bool, absl::optional<std::__1::reference_wrapper<Envoy::ProcessObject> >, bool, bool, unsigned int)::$_0::operator()() const /proc/self/cwd/test/integration/server.cc:92:9 (listener_lds_integration_test+0x26ef020)
2020-03-25T09:30:33.5136604Z     #41 decltype(std::__1::forward<Envoy::IntegrationTestServer::start(Envoy::Network::Address::IpVersion, std::__1::function<void ()>, bool, bool, absl::optional<std::__1::reference_wrapper<Envoy::ProcessObject> >, bool, bool, unsigned int)::$_0&>(fp)()) std::__1::__invoke<Envoy::IntegrationTestServer::start(Envoy::Network::Address::IpVersion, std::__1::function<void ()>, bool, bool, absl::optional<std::__1::reference_wrapper<Envoy::ProcessObject> >, bool, bool, unsigned int)::$_0&>(Envoy::IntegrationTestServer::start(Envoy::Network::Address::IpVersion, std::__1::function<void ()>, bool, bool, absl::optional<std::__1::reference_wrapper<Envoy::ProcessObject> >, bool, bool, unsigned int)::$_0&) /opt/llvm/bin/../include/c++/v1/type_traits:3530:1 (listener_lds_integration_test+0x26eee90)
2020-03-25T09:30:33.5140022Z     #42 void std::__1::__invoke_void_return_wrapper<void>::__call<Envoy::IntegrationTestServer::start(Envoy::Network::Address::IpVersion, std::__1::function<void ()>, bool, bool, absl::optional<std::__1::reference_wrapper<Envoy::ProcessObject> >, bool, bool, unsigned int)::$_0&>(Envoy::IntegrationTestServer::start(Envoy::Network::Address::IpVersion, std::__1::function<void ()>, bool, bool, absl::optional<std::__1::reference_wrapper<Envoy::ProcessObject> >, bool, bool, unsigned int)::$_0&) /opt/llvm/bin/../include/c++/v1/__functional_base:348:9 (listener_lds_integration_test+0x26eedf0)
2020-03-25T09:30:33.5143121Z     #43 std::__1::__function::__alloc_func<Envoy::IntegrationTestServer::start(Envoy::Network::Address::IpVersion, std::__1::function<void ()>, bool, bool, absl::optional<std::__1::reference_wrapper<Envoy::ProcessObject> >, bool, bool, unsigned int)::$_0, std::__1::allocator<Envoy::IntegrationTestServer::start(Envoy::Network::Address::IpVersion, std::__1::function<void ()>, bool, bool, absl::optional<std::__1::reference_wrapper<Envoy::ProcessObject> >, bool, bool, unsigned int)::$_0>, void ()>::operator()() /opt/llvm/bin/../include/c++/v1/functional:1533:16 (listener_lds_integration_test+0x26eed90)
2020-03-25T09:30:33.5146142Z     #44 std::__1::__function::__func<Envoy::IntegrationTestServer::start(Envoy::Network::Address::IpVersion, std::__1::function<void ()>, bool, bool, absl::optional<std::__1::reference_wrapper<Envoy::ProcessObject> >, bool, bool, unsigned int)::$_0, std::__1::allocator<Envoy::IntegrationTestServer::start(Envoy::Network::Address::IpVersion, std::__1::function<void ()>, bool, bool, absl::optional<std::__1::reference_wrapper<Envoy::ProcessObject> >, bool, bool, unsigned int)::$_0>, void ()>::operator()() /opt/llvm/bin/../include/c++/v1/functional:1707:12 (listener_lds_integration_test+0x26ed45f)
2020-03-25T09:30:33.5148190Z     #45 std::__1::__function::__value_func<void ()>::operator()() const /opt/llvm/bin/../include/c++/v1/functional:1860:16 (listener_lds_integration_test+0x25c6826)
2020-03-25T09:30:33.5149178Z     #46 std::__1::function<void ()>::operator()() const /opt/llvm/bin/../include/c++/v1/functional:2419:12 (listener_lds_integration_test+0x25c64b8)
2020-03-25T09:30:33.5150257Z     #47 Envoy::Thread::ThreadImplPosix::ThreadImplPosix(std::__1::function<void ()>)::$_0::operator()(void*) const /proc/self/cwd/source/common/common/posix/thread_impl.cc:33:9 (listener_lds_integration_test+0x5663c9b)
2020-03-25T09:30:33.5151502Z     #48 Envoy::Thread::ThreadImplPosix::ThreadImplPosix(std::__1::function<void ()>)::$_0::__invoke(void*) /proc/self/cwd/source/common/common/posix/thread_impl.cc:32:7 (listener_lds_integration_test+0x5663c28)
2020-03-25T09:30:33.5152588Z 
2020-03-25T09:30:33.5152950Z   Previous read of size 8 at 0x7b480001ffa8 by main thread:
2020-03-25T09:30:33.5154235Z     #0 std::__1::__list_imp<std::__1::unique_ptr<Envoy::Server::ListenerImpl, std::__1::default_delete<Envoy::Server::ListenerImpl> >, std::__1::allocator<std::__1::unique_ptr<Envoy::Server::ListenerImpl, std::__1::default_delete<Envoy::Server::ListenerImpl> > > >::begin() /opt/llvm/bin/../include/c++/v1/list:618:32 (listener_lds_integration_test+0x2a2ad19)
2020-03-25T09:30:33.5156157Z     #1 std::__1::list<std::__1::unique_ptr<Envoy::Server::ListenerImpl, std::__1::default_delete<Envoy::Server::ListenerImpl> >, std::__1::allocator<std::__1::unique_ptr<Envoy::Server::ListenerImpl, std::__1::default_delete<Envoy::Server::ListenerImpl> > > >::begin() /opt/llvm/bin/../include/c++/v1/list:944:59 (listener_lds_integration_test+0x29ee9bb)
2020-03-25T09:30:33.5157605Z     #2 Envoy::Server::ListenerManagerImpl::listeners() /proc/self/cwd/source/server/listener_manager_impl.cc:595:29 (listener_lds_integration_test+0x29cc721)
2020-03-25T09:30:33.5159083Z     #3 Envoy::BaseIntegrationTest::registerTestServerPorts(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) /proc/self/cwd/test/integration/integration.cc:414:61 (listener_lds_integration_test+0x2646ed3)
2020-03-25T09:30:33.5160697Z     #4 Envoy::(anonymous namespace)::ListenerIntegrationTest_RemoveLastUninitializedListener_Test::TestBody() /proc/self/cwd/test/integration/listener_lds_integration_test.cc:198:3 (listener_lds_integration_test+0x242bf87)
2020-03-25T09:30:33.5162948Z     #5 non-virtual thunk to Envoy::(anonymous namespace)::ListenerIntegrationTest_RemoveLastUninitializedListener_Test::TestBody() /proc/self/cwd/test/integration/listener_lds_integration_test.cc (listener_lds_integration_test+0x242d0f2)
2020-03-25T09:30:33.5164371Z     #6 void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2433:10 (listener_lds_integration_test+0x70dafec)
2020-03-25T09:30:33.5165824Z     #7 void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2469:14 (listener_lds_integration_test+0x70bf6ee)
2020-03-25T09:30:33.5167080Z     #8 testing::Test::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2508:5 (listener_lds_integration_test+0x70a47d1)
2020-03-25T09:30:33.5167941Z     #9 testing::TestInfo::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2684:11 (listener_lds_integration_test+0x70a5673)
2020-03-25T09:30:33.5168809Z     #10 testing::TestSuite::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2816:28 (listener_lds_integration_test+0x70a60aa)
2020-03-25T09:30:33.5169722Z     #11 testing::internal::UnitTestImpl::RunAllTests() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:5338:44 (listener_lds_integration_test+0x70b584d)
2020-03-25T09:30:33.5171224Z     #12 bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2433:10 (listener_lds_integration_test+0x70e1f9c)
2020-03-25T09:30:33.5172908Z     #13 bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2469:14 (listener_lds_integration_test+0x70c35be)
2020-03-25T09:30:33.5174217Z     #14 testing::UnitTest::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:4925:10 (listener_lds_integration_test+0x70b510b)
2020-03-25T09:30:33.5175207Z     #15 RUN_ALL_TESTS() /proc/self/cwd/external/com_google_googletest/googletest/include/gtest/gtest.h:2473:46 (listener_lds_integration_test+0x4e5a367)
2020-03-25T09:30:33.5176084Z     #16 Envoy::TestRunner::RunTests(int, char**) /proc/self/cwd/test/test_runner.cc:134:10 (listener_lds_integration_test+0x4e59817)
2020-03-25T09:30:33.5176861Z     #17 main /proc/self/cwd/test/main.cc:46:10 (listener_lds_integration_test+0x4e5676a)
```

**Comment by lambdai** [comment]
macos test is failing elsewhere and it's not related to this PR

File: test/integration/integration.cc [hunk 1]
```diff
@@ -414,10 +414,25 @@ void BaseIntegrationTest::setUpstreamAddress(
...
+    listeners = test_server_->server().listenerManager().listeners();
+    l.Lock();
+    listeners_ready = true;
+    l.Unlock();
+  });
+  l.LockWhen(absl::Condition(read_flag, &listeners_ready));
+  l.Unlock();
+
   auto listener_it = listeners.cbegin();
+  auto port_it = port_names.cbegin();
```
Comment (htuch) [inline test/integration/integration.cc:435]: Wondering if this is safe given the comment about call stack validity in https://github.com/envoyproxy/envoy/blob/master/include/envoy/server/listener_manager.h#L164.

Comment (lambdai) [reply test/integration/integration.cc:435]: I agree the production should not violate the rules. 
But for an integration test it's probably good enough. 
Well, if the PR is not strong enough, the next step is to provide a callback and copy the needed info in master thread.

File: test/integration/integration.cc [hunk 2]
```diff
@@ -414,10 +414,25 @@ void BaseIntegrationTest::setUpstreamAddress(
...
+  bool listeners_ready = false;
+  absl::Mutex l;
+  std::vector<std::reference_wrapper<Network::ListenerConfig>> listeners;
+  test_server_->server().dispatcher().post([this, &listeners, &listeners_ready, &l]() {
+    listeners = test_server_->server().listenerManager().listeners();
+    l.Lock();
+    listeners_ready = true;
+    l.Unlock();
+  });
+  l.LockWhen(absl::Condition(read_flag, &listeners_ready));
```
Comment (htuch) [inline test/integration/integration.cc:431]: Should a `CondVar` be used here, rather than manually doing this with a bool?

Comment (lambdai) [reply test/integration/integration.cc:431]: Interesting... My understanding is that CondVar doesn't maintain the bool flag in this scenario.
It's always 3 components: Mutex, CondVar and Flag
and the core is 
```
while (flag) {
   cond_var.wait(mutex);
}
```
Not sure how we could save the bool flag...

Comment (lambdai) [reply test/integration/integration.cc]: While looking up CondVar I accidentally find read_flag is a overhead. Removed

**Pull request merged**